app/models/population.rb

```diff
@@ -0,0 +1,56 @@
+class Population < ApplicationRecord
+  belongs_to :location
+
+  validates :date, presence: true
+  validates :operational_capacity, presence: true, numericality: { greater_than: 0 }
+  validates :usable_capacity, presence: true, numericality: { greater_than: 0 }
+  validates :unlock, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :bedwatch, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :overnights_in, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :overnights_out, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :out_of_area_courts, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :discharges, presence: true, numericality: { greater_than_or_equal_to: 0 }
+
+  def moves_from
+    location.moves_from.prison_transfer.not_cancelled.where(date: date)
+  end
+
+  def moves_to
+    location.moves_to.prison_transfer.not_cancelled.where(date: date)
+  end
+
+  def free_spaces
+    usable_capacity - unlock - bedwatch - overnights_in - moves_to.count + overnights_out + out_of_area_courts + discharges + moves_from.count
+  end
+
+  def self.free_spaces_date_range(locations, date_range)
+    {}.tap do |locations_hash|
+      date_range.each do |date|
+        free_spaces_for_date(locations, date).each do |row|
```

Comment:
Is it possible/desirable to push the grouping by date into SQL so we only have one db query? I'm also wondering whether we're at a point where properly formatted plain SQL would be easier to read?

Reply:
I explored this but the tradeoff is a massive increase in DB query complexity. This is just about acceptable from a complexity point of view and results in one query per date. The front end queries 5 days ahead, so that's 5 days max. I'm been testing locally with a reasonable volume of data copied from staging and seeing really good performance - returning all of the prisons about once takes around 300ms - a page of 20 at at time is more like 40ms. So I'm happy with the tradeoff here.

Reply:
Okay. Makes sense.